app/src/main/java/com/android/unio/ui/map/Map.kt

```diff
@@ -16,35 +23,115 @@ import androidx.compose.runtime.Composable
 import com.google.maps.android.compose.rememberCameraPositionState
 import java.util.Calendar
 import java.util.concurrent.TimeUnit
 
+val EPFL_COORDINATES = LatLng(46.518831258, 6.559331096)
+const val APPROXIMATE_CIRCLE_RADIUS = 30.0
+const val APPROXIMATE_CIRCLE_OUTLINE_WIDTH = 2f
+
 @OptIn(ExperimentalMaterial3Api::class)
 @Composable
 fun MapScreen(
     navigationAction: NavigationAction,
     eventViewModel: EventViewModel,
-    userViewModel: UserViewModel
+    userViewModel: UserViewModel,
+    mapViewModel: MapViewModel
 ) {
   val context = LocalContext.current
+  val cameraPositionState = rememberCameraPositionState()
+  val userLocation by mapViewModel.userLocation.collectAsState()
+  var initialCentered by remember { mutableStateOf(false) }
+  var isMyLocationEnabled by remember { mutableStateOf(false) }
+  var showApproximateCircle by remember { mutableStateOf(false) }
+
+  val permissionLauncher =
+      rememberLauncherForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) {
+          permissions ->
+        when {
+          permissions.getOrDefault(Manifest.permission.ACCESS_FINE_LOCATION, false) -> {
+            mapViewModel.fetchUserLocation(context)
+            isMyLocationEnabled = true
+            showApproximateCircle = false
+          }
+          permissions.getOrDefault(Manifest.permission.ACCESS_COARSE_LOCATION, false) -> {
+            mapViewModel.fetchUserLocation(context)
+            isMyLocationEnabled = false
+            showApproximateCircle = true
+          }
+          else -> {
+            Log.e("MapScreen", "Location permission is not granted.")
+          }
+        }
+      }
+
+  /** Request location permissions. */
+  fun requestPermissions() {
+    permissionLauncher.launch(
+        arrayOf(
+            Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION))
+  }
+
+  // Check what permissions are already granted
+  LaunchedEffect(Unit) {
+    when (PackageManager.PERMISSION_GRANTED) {
+      ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) -> {
+        isMyLocationEnabled = true
+        showApproximateCircle = false
+        mapViewModel.fetchUserLocation(context)
+      }
+      ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_COARSE_LOCATION) -> {
+        isMyLocationEnabled = false
+        showApproximateCircle = true
+        requestPermissions()
+        mapViewModel.fetchUserLocation(context)
+      }
+      else -> {
+        requestPermissions()
+      }
+    }
+    cameraPositionState.position =
+        CameraPosition.fromLatLngZoom(userLocation ?: EPFL_COORDINATES, 10f)
+  }
+
+  // Center map on the user's location initially if available
+  LaunchedEffect(userLocation) {
+    if (userLocation != null && !initialCentered) {
+      cameraPositionState.position = CameraPosition.fromLatLngZoom(userLocation!!, 15f)
+      initialCentered = true
+    }
+  }
```

Comment:
For now this is redundant, as userLocation is only updated when requesting permissions.

Reply:
You're right, but it won't be when periodic updates of the view model are introduced next week.